src/raw_node.rs

```diff
@@ -174,18 +176,50 @@ impl Ready {
     /// If it contains a MsgSnap message, the application MUST report back to raft
     /// when the snapshot has been received or has failed by calling ReportSnapshot.
     #[inline]
-    pub fn messages(&self) -> &Vec<Vec<Message>> {
-        self.light.messages()
+    pub fn messages(&self) -> &[Message] {
+        if !self.is_persisted_msg {
+            self.light.messages()
+        } else {
+            &[]
+        }
     }
 
     /// Take the Messages.
     #[inline]
-    pub fn take_messages(&mut self) -> Vec<Vec<Message>> {
-        self.light.take_messages()
+    pub fn take_messages(&mut self) -> Vec<Message> {
+        if !self.is_persisted_msg {
+            self.light.take_messages()
+        } else {
+            Vec::new()
+        }
+    }
+
+    /// Persisted Messages specifies outbound messages to be sent AFTER the HardState,
+    /// Entries and Snapshot are persisted to stable storage.
+    #[inline]
+    pub fn persisted_messages(&self) -> &[Message] {
+        if self.is_persisted_msg {
+            self.light.messages()
+        } else {
+            &[]
+        }
+    }
+
+    /// Take the Persisted Messages.
+    #[inline]
+    pub fn take_persisted_messages(&mut self) -> Vec<Message> {
+        if self.is_persisted_msg {
+            self.light.take_messages()
+        } else {
+            Vec::new()
+        }
     }
 
-    /// MustSync indicates whether the HardState and Entries must be synchronously
-    /// written to disk or if an asynchronous write is permissible.
+    /// MustSync is true if and only if
+    /// 1. no HardState or only its commit is different from before
+    /// 2. no Entries and Snapshot
```

Comment:
Seems it's different from the implementation?

Reply:
My mistake. The `true` should be `false`.

Reply:
My mistake. The `true` should be `false`.